go/runtime/scheduling/simple/txpool/priorityqueue/priority_queue.go

```diff
@@ -0,0 +1,218 @@
+	q.priorityIndex.ReplaceOrInsert(item)
+	q.transactions[tx.Hash()] = item
+
+	return nil
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) GetBatch(force bool) []*transaction.CheckedTransaction {
+	q.Lock()
+	defer q.Unlock()
+
+	// Check if a batch is ready.
+	// XXX: could also check if any other weight limits are reached.
+	if uint64(q.priorityIndex.Len()) < q.weightLimits[transaction.WeightCount] && !force {
+		return nil
+	}
+
+	var batch []*transaction.CheckedTransaction
+	batchWeights := make(map[string]uint64)
+	for w := range q.weightLimits {
+		batchWeights[w] = 0
+	}
+	toRemove := []*item{}
+	q.priorityIndex.Ascend(func(i btree.Item) bool {
+		item := i.(*item)
+
+		// Check if the call fits into the batch.
+		// XXX: potentially there could be smaller transactions that would
+		// fit, which this will miss. Could do some lookahead.
+		for w, limit := range q.weightLimits {
+			batchWeight := batchWeights[w]
+
+			txW := item.tx.Weight(w)
+			// Transaction weight greater than the limit. Drop the tx from the pool.
+			if txW > limit {
+				toRemove = append(toRemove, item)
+				return true
+			}
+
+			// Batch full, schedule the batch.
+			if batchWeight+txW > limit {
+				return false
+			}
+		}
+
+		// Add the tx to the batch.
+		batch = append(batch, item.tx)
+		for w, val := range item.tx.Weights() {
+			if _, ok := batchWeights[w]; ok {
+				batchWeights[w] += val
+			}
+		}
+
+		return true
+	})
+
+	// Remove transaction discovered to be to big to even fit the batch.
+	// This can happen if weight limits changed, after the transaction was
+	// already set to be scheduled.
+	for _, tw := range toRemove {
+		delete(q.transactions, tw.tx.Hash())
+		q.priorityIndex.Delete(tw)
+	}
+
+	return batch
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) RemoveBatch(batch []hash.Hash) error {
+	q.Lock()
+	defer q.Unlock()
+
+	for _, txHash := range batch {
+		if item, ok := q.transactions[txHash]; ok {
+			q.priorityIndex.Delete(item)
+			delete(q.transactions, txHash)
+		}
+	}
+	if len(q.transactions) != q.priorityIndex.Len() {
+		panic(fmt.Errorf("inconsistent sizes of the underlying list (%v) and map (%v) after RemoveBatch", q.priorityIndex.Len(), len(q.transactions)))
+	}
+
+	return nil
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) IsQueued(txHash hash.Hash) bool {
+	q.Lock()
+	defer q.Unlock()
+
+	return q.isQueuedLocked(txHash)
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) Size() uint64 {
+	q.Lock()
+	defer q.Unlock()
+
+	return uint64(q.priorityIndex.Len())
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) UpdateConfig(cfg api.Config) error {
+	q.Lock()
+	defer q.Unlock()
+	q.maxTxPoolSize = cfg.MaxPoolSize
+	q.weightLimits = cfg.WeightLimits
+
+	// Any transaction not within the new limits will get removed during GetBatch iteration.
+
+	return nil
+}
+
+// Implements api.TxPool.
+func (q *priorityQueue) Clear() {
+	q.Lock()
+	defer q.Unlock()
+
+	q.priorityIndex = btree.New(2)
```

Comment:
Should probably do `q.priorityIndex.Clear` instead:
> This can be much faster than calling Delete on all elements, because that requires finding/removing each element in the tree and updating the tree accordingly. It also is somewhat faster than creating a new tree to replace the old one, because nodes from the old tree are reclaimed into the freelist for use by the new one, instead of being lost to the garbage collector.